components/renku_data_services/k8s_watcher/db.py

```diff
@@ -0,0 +1,319 @@
+            session.add(obj_orm)
+            return
+
+    async def delete(self, meta: K8sObjectMeta) -> None:
+        """Delete an object from the cache."""
+        async with self.__session_maker() as session, session.begin():
+            obj_orm = await self.__get(meta, session)
+            if obj_orm is None:
+                return
+            await session.delete(obj_orm)
+            return
+
+    async def get(self, meta: K8sObjectMeta) -> K8sObject | None:
+        """Get a single object from the cache."""
+        async with self.__session_maker() as session, session.begin():
+            obj = await self.__get(meta, session)
+            if not obj:
+                return None
+            return meta.with_manifest(obj.manifest)
+
+    async def list(self, filter: ListFilter) -> AsyncIterable[K8sObject]:
+        """List objects from the cache."""
+        async with self.__session_maker() as session, session.begin():
+            stmt = select(K8sObjectORM)
+            if filter.name:
+                stmt = stmt.where(K8sObjectORM.name == filter.name)
+            if filter.namespace:
+                stmt = stmt.where(K8sObjectORM.namespace == filter.namespace)
+            if filter.cluster:
+                stmt = stmt.where(K8sObjectORM.cluster == filter.cluster)
+            if filter.kind:
+                stmt = stmt.where(K8sObjectORM.kind == filter.kind)
+            if filter.version:
+                stmt = stmt.where(K8sObjectORM.version == filter.version)
+            if filter.label_selector:
+                stmt = stmt.where(
+                    sqlalchemy.text("manifest -> 'metadata' -> 'labels' @> :labels").bindparams(
+                        labels=json.dumps(filter.label_selector)
+                    )
+                )
+            async for res in await session.stream_scalars(stmt):
+                yield res.dump()
+
+
+class CachedK8sClient(K8sClient):
+    """A wrapper around a kr8s k8s client."""
+
+    def __init__(self, clusters: dict[ClusterId, Cluster], cache: K8sCache) -> None:
+        super().__init__(clusters)
+        self.__cache = cache
+
+    async def create(self, obj: K8sObject) -> K8sObject:
+        """Create the k8s object."""
+        obj = await super().create(obj)
+        await self.__cache.upsert(obj)
+        return obj
+
+    async def patch(self, meta: K8sObjectMeta, patch: dict[str, Any]) -> K8sObject:
+        """Patch a k8s object."""
+        obj = await super().patch(meta, patch)
+        await self.__cache.upsert(obj)
+        return obj
+
+    async def delete(self, meta: K8sObjectMeta) -> None:
+        """Delete a k8s object."""
+        await super().delete(meta)
+        await self.__cache.delete(meta)
+
+    async def get(self, meta: K8sObjectMeta) -> K8sObject | None:
+        """Get a specific k8s object, None is returned if the object does not exist."""
+        res = await self.__cache.get(meta)
+        if res is None:
+            return None
```

Comment:
IMHO here we should try to go to the cluster, and if we get a result update the cache. then return the result:
```
if res is None:
  # Let's try to retrieve it directly from the cluster, if there is an object, cache it.
  obj = await super().get(meta)
  if obj is not None:
    await self.__cache.upsert(obj)
  return obj
```

Reply:
Yeah this can be an optional thing we do. If the client is configured to be allowed to skip the cache. In production we cannot do this because we will take down the k8s api. But we can add the configuration/option to do so.